### ISSUE TITLE
Saving and loading TransformerDecoder/Encoder with custom objects rebuilds the layer.

### ISSUE THREAD

**Issue by jessechancy** [keras-team/keras-nlp#219]
**Describe the bug**
When loading a model with TransformerDecoder, with the custom objects option, it rebuilds and reinitializes the layers within TransformerDecoder.

**To Reproduce**
In keras_nlp/layers/transformer_decoder_test, change the test case of test_save_model by changing the load_model line  to:
(https://github.com/keras-team/keras-nlp/blob/38ca02b96b1a9100bcc76d123b399ecbc1f902c2/keras_nlp/layers/transformer_decoder_test.py#L194

loaded_model = keras.models.load_model(path, custom_objects={"TransformerDecoder": decoder})

After that, run the following command to run the test:

python3 -m pytest keras_nlp/layers/transformer_decoder_test.py

This will result in a mismatch error in test_save_model.

**Expected behavior**
Layers within TransformerDecoder should not be reinitialized, the saved weights should be kept.

**Comment by elbamos** [comment]
I am experiencing an issue with a model using TransformerEncoder blocks, where when the model is saved and reloaded the accuracy drops so far it appears to be an untrained model. Is that the expected result of this bug? 

If so, is there any workaround or solution? I have a model that works very very well which I really would like to be deploying to production.

**Comment by jbischof** [comment]
@elbamos can you provide us with a minimal example colab reproducing the issue? Thanks!

**Comment by elbamos** [comment]
Would you need a colab - could you take a jupyter notebook?

**Comment by jbischof** [comment]
You can open a jupyter notebook in Colab so hopefully not an issue. Colabs are much easier to share

**Comment by elbamos** [comment]
Thanks, I'll come up with something in the next day or two.

**Comment by LazurasLong** [comment]
I encountered something similar.

Upon saving the initial model i was given this warning:

WARNING:absl:Found untraced functions such as
token_embedding2_layer_call_fn,
token_embedding2_layer_call_and_return_conditional_losses,
position_embedding2_layer_call_fn,
position_embedding2_layer_call_and_return_conditional_losses,
multi_head_attention_layer_call_fn while saving (showing 5 of 56).
These functions will not be directly callable after loading.


From what I can tell this is becuase I I haven't gone back and retested
using get_config to see if the issue is related to having nominally custom
layers

**Comment by mattdangerw** [comment]
Thanks! Reading over everything here. I actually think the original issue here is no longer valid. The code it is linking to is very old. For the decoder issues described at the top, two things are no longer valid...

 - There should be no need to pass custom objects during save and load. All our objects are registered with Keras during library import, so they will automatically be found and restored during saving/loading.
 - We now should be correctly restoring weights for decoder layers on save/loading. There a few lines for that [here](https://github.com/keras-team/keras-nlp/blob/c21f852426644e08631fcc1032059d55c7142ef6/keras_nlp/layers/transformer_decoder.py#L113-L116).

@LazurasLong and @elbamos the issues you are describing sound unrelated to this. It might be best to track these in new issue, with a colab to reproduce the issue. Overall saving and loading is quite complex, so worth looking into all this.

I will mark this issue as closed to reduce confusion on the original issue, but please open new ones for any saving problems.
